x-pack/elastic-agent/pkg/core/plugin/process/app.go

```diff
@@ -192,16 +196,32 @@ func (a *Application) watch(ctx context.Context, p app.Taggable, proc *process.I
 		case ps := <-a.waitProc(proc.Process):
 			procState = ps
 		case <-a.bgContext.Done():
-			a.Stop()
+			return
+		case <-ctx.Done():
+			// closer called
 			return
 		}
 
 		a.appLock.Lock()
 		if a.state.ProcessInfo != proc {
+			// kill original process if possible
+			if proc != nil && proc.Process != nil {
+				_ = proc.Process.Kill()
+			}
+
 			// already another process started, another watcher is watching instead
 			a.appLock.Unlock()
 			return
 		}
+
+		// stop the watcher
+		a.stopWatcher(a.state.ProcessInfo)
+
+		// was already stopped by Stop, do not restart
+		if a.state.Status == state.Stopped {
+			return
```

Comment:
are we missing a `a.appLock.Unlock()t here?

Reply:
true

Reply:
checking the complete function body, we always use `Unlock` for each exit path. Better add the `defer unlock` right after the `lock` in order to reduce the chance of introducing new deadlocks in the future.